qiskit/circuit/library/generalized_gates/permutation.py

```diff
@@ -16,20 +16,20 @@
 
 import numpy as np
 
-from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.quantumcircuit import Gate
 from qiskit.circuit.exceptions import CircuitError
 
 
-class Permutation(QuantumCircuit):
-    """An n_qubit circuit that permutes qubits."""
+class Permutation(Gate):
```

Comment:
I'm a bit concerned about making this change from a backwards compatibility PoV in a single release. While I agree with this being a `Gate` subclass instead of a `QuantumCircuit` is where we want to be I'm worried about any users of this class today that may be using it with a  circuit as is. For example the docs here show a good example of this with `Permutation(...).draw()`, but there are others I can think of the top of my head would be something like: `Permutation(...).compose(sub_circuit)` or `Permutation.measure_all()` all of which were previously completely valid which would be broken by this change.

 I'm thinking we need to figure out a better way to manage this migration which leaves the `QuantumCircuit` based `Permutation` class in place as is for now but establishes a new `Gate` based class as a standalone thing (like `PermutationGate` or something) in this PR and maybe update this `QuantumCircuit` based one to just wrap the new `Gate` class. Then in the next release we can deprecate the `QuantumCircuit` version and tell people to use the gate directly.

Reply:
Thanks, this is a very good point. I keep relying on qiskit-terra tests to make sure that nothing gets broken, yet I keep forgetting that there is also a lot of other code outside of qiskit-terra. The example with `Permutation.measure_all()` clearly shows why we need to keep the old functionality. It should be quite straightforward for me to keep the old implementation and to rename the old implementation, in this way this development no longer changes any existing behavior, it only adds new things.

I have already started doing changes, but here are a couple of preliminary thoughts. The name `PermutationGate` for the new class is good. For objects of type `PermutationGate` we can still have `op.name = "permutation"`, since there is no danger to confuse permutation circuits and permutation gates (and in any case the name assigned to a permutation circuit is of the form `permutation_[2,4,3,0,1]`). In particular, we can keep using names like `permutation.kms` in `entry_points` to specify permutation synthesis algorithms for `PermutationGate`.

The suggested deprecation flow also allows to avoid passing `num_qubits` and `seed` to the initializer of the `PermutationGate`, making the code a bit cleaner. There is a function `np.random.permutation` that returns a random permutation pattern, from which a `PermutationGate` can be constructed.